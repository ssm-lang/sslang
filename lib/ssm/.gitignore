# Prerequisites
*.d

# Object files
*.o
*.ko
*.obj
*.elf
counter
counter2
clock
fib
fib2
fib3
onetwo

# Linker output
*.ilk
*.map
*.exp

# Precompiled Headers
*.gch
*.pch

# Libraries
*.lib
*.a
*.la
*.lo

# Shared objects (inc. Windows DLLs)
*.dll
*.so
*.so.*
*.dylib

# Executables
*.exe
*.app
*.i*86
*.x86_64
*.hex

# Debug files
*.dSYM/
*.su
*.idb
*.pdb

# Kernel Module Compile Results
*.mod*
*.cmd
.tmp_versions/
modules.order
Module.symvers
Mkfile.old
dkms.conf

# gcov output
*.gcov
*.gcda
*.gcno

build/

.pio

compile_commands.json

vgcore.*
