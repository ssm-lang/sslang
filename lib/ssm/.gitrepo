; DO NOT EDIT (unless you know what you are doing)
;
; This subdirectory is a git "subrepo", and this file is maintained by the
; git-subrepo command. See https://github.com/git-commands/git-subrepo#readme
;
[subrepo]
	remote = ssm-runtime
	branch = main
	commit = a9bf5272efed87994389613dd29e8f99bb7465c5
	parent = bfbde4f3b1727f9a6ffe768391f10c50e3f2fbdb
	method = merge
	cmdver = 0.4.3
