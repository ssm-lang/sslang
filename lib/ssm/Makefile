######################
# SSM runtime Makefile
######################
#
# For a list of useful available targets, run `make help'.
#
# Note that this Makefile is only used to build and test the platform-generic
# core of the SSM runtime; platform-specific builds may use their own build
# system that ignores this Makefile.
#
# Any .c files under EXE_DIR are linked against this library to produce an
# executable.
#
# This Makefile also a version of the runtime library instrumented for code
# coverage analysis. The build artifacts for instrumented code are mangled using
# the test_ prefix. This version of the runtime library is used to link against
# .c files in TEST_DIR.
#
# To actually run tests, please see the runtests.sh script.

LIB_NAME := ssm

MAKE_CC ?= cc
MAKE_LD ?= $(MAKE_CC)
MAKE_AR ?= ar

PLATFORM ?= simulation

BUILD_DIR := build
SRC_DIR := src
INC_DIR := include
TEST_DIR := test
EXE_DIR := examples
DOC_DIR := doc
PLATFORM_DIR := platform/$(PLATFORM)
GCOVR_CFG := test/gcovr.cfg

LIB_SRC := $(wildcard $(SRC_DIR)/*.c)
LIB_INC := $(wildcard include/*.h)

LIB_OBJ := $(patsubst $(SRC_DIR)/%.c, $(BUILD_DIR)/%.o, $(LIB_SRC))
LIB_TGT := $(BUILD_DIR)/lib$(LIB_NAME).a

DOC_CFG := doc/Doxyfile
DOC_SRC := README.md $(wildcard doc/*.md doc/*.dox) $(LIB_SRC) $(LIB_INC)
DOC_TGT := $(BUILD_DIR)/doc

EXE_SRC := $(wildcard $(EXE_DIR)/*.c)
EXE_OBJ := $(patsubst $(EXE_DIR)/%.c, $(BUILD_DIR)/%.o, $(EXE_SRC))
EXE_TGT := $(patsubst %.o, %, $(EXE_OBJ))
EXE_INC := $(wildcard $(EXE_DIR)/*.h)

TLIB_NAME := t$(LIB_NAME)
TLIB_OBJ := $(patsubst $(SRC_DIR)/%.c, $(BUILD_DIR)/test_%.o, $(LIB_SRC))
TLIB_TGT := $(BUILD_DIR)/lib$(TLIB_NAME).a

TEST_SRC := $(wildcard $(TEST_DIR)/*.c)
TEST_OBJ := $(patsubst $(TEST_DIR)/%.c, $(BUILD_DIR)/test_%.o, $(TEST_SRC))
TEST_TGT := $(patsubst %.o, %, $(TEST_OBJ))

PLATFORM_SRC := $(wildcard $(PLATFORM_DIR)/src/*.c)
PLATFORM_OBJ := $(patsubst $(PLATFORM_DIR)/src/%.c, $(BUILD_DIR)/%.o, $(PLATFORM_SRC))
$(info PLATFORM is $(PLATFORM))

COV_TGT := $(BUILD_DIR)/coverage.xml

CC = $(MAKE_CC)
CFLAGS += -g -I$(INC_DIR) -O -Wall -pedantic -DSSM_TIMER64_PRESENT # -std=c99
# C99 removes access to some POSIX high-precision timing APIs
# TODO: only use C99 for building core lib

# Check whether valgrind is available.
ifeq ($(shell command -v valgrind),)
# If not, we probably shouldn't try to include <valgrind/valgrind.h>,
# which we use to instrument our memory allocator.
# 
# NOTE: this is just a heuristic that may not be very reliable.
# We should probably implement better dependency management.
CFLAGS += -DNVALGRIND
endif

TEST_CFLAGS = $(CFLAGS) -g -DSSM_DEBUG --coverage

LD = $(MAKE_LD)
LDFLAGS = -L$(BUILD_DIR)
TEST_LDFLAGS = $(LDFLAGS) --coverage
LDLIBS= -lpthread

AR = $(MAKE_AR)
ARFLAGS = -cr

PHONY += lib exes tests cov docs
lib: $(LIB_TGT)
exes: $(EXE_TGT)
tests: $(TEST_TGT)
cov: $(COV_TGT)
docs: $(DOC_TGT)

$(LIB_TGT): $(LIB_OBJ)
	rm -f $@
	$(AR) $(ARFLAGS) $@ $+

$(TLIB_TGT): $(TLIB_OBJ)
	rm -f $@
	$(AR) $(ARFLAGS) $@ $+

$(EXE_TGT): %: %.o $(PLATFORM_OBJ) $(LIB_TGT)
	$(LD) $(LDFLAGS) -o $@ $@.o $(PLATFORM_OBJ) -l$(LIB_NAME) $(LDLIBS)

$(TEST_TGT): %: %.o $(TLIB_TGT)
	$(LD) $(TEST_LDFLAGS) -o $@ $@.o -l$(TLIB_NAME) $(LDLIBS)

vpath %.c $(SRC_DIR) $(EXE_DIR) $(TEST_DIR) $(PLATFORM_DIR)/src

$(EXE_OBJ): $(BUILD_DIR)/%.o: %.c $(LIB_INC) $(EXE_INC) | $(BUILD_DIR)
	$(CC) $(CFLAGS) -c -o $@ $<

$(LIB_OBJ): $(BUILD_DIR)/%.o: %.c $(LIB_INC) | $(BUILD_DIR)
	$(CC) $(CFLAGS) -c -o $@ $<

$(TLIB_OBJ) $(TEST_OBJ): $(BUILD_DIR)/test_%.o: %.c $(LIB_INC) | $(BUILD_DIR)
	rm -f $(patsubst %.o, %.gcda, $@) $(patsubst %.o, %.gcno, $@)
	$(CC) $(TEST_CFLAGS) -c -o $@ $<

$(PLATFORM_OBJ): $(BUILD_DIR)/%.o: %.c $(LIB_INC) | $(BUILD_DIR)
	$(CC) $(CFLAGS) -c -o $@ $<

$(COV_TGT): $(GCOVR_CFG) $(TEST_TGT)
	@for i in $(TEST_TGT) ; do \
		echo ./$$i ;\
		./$$i >/dev/null || exit $$? ;\
	done
	gcovr --config $(GCOVR_CFG)

$(DOC_TGT): $(DOC_CFG) $(DOC_SRC) | $(BUILD_DIR)
	doxygen doc/Doxyfile

$(BUILD_DIR):
	mkdir -p $@

PHONY += clean
clean:
	rm -rf build

PHONY += help
help:
	@echo "Available phony targets:" $(PHONY)
	@echo
	@echo "lib     Build SSM runtime library     [build/libssm.a]"
	@echo "exes    Build programs in examples/   [build/<example-name>]"
	@echo "tests   Build tests in test/          [build/test_<test-name>]"
	@echo "cov     Build coverage report         [build/<filename.c>.gcov] (depends on gcovr)"
	@echo "docs    Build code documentation      [build/doc]               (depends on doxygen)"
	@echo "clean   Remove build directory"
	@echo "help    Show this help menu"
	@echo
	@echo "Available example targets:" $(EXE_TGT)
	@echo "Available test targets:" $(TEST_TGT)

.PHONY: $(PHONY)
