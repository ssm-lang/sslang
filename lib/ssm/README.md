# The Sparse Synchronous Model Runtime

<div>
<a>
  <img src="https://github.com/ssm-lang/ssm-runtime/actions/workflows/test.yml/badge.svg?branch=dev">
</a>
<a href="https://ssm-lang.github.io/ssm-runtime">
  <img src="https://github.com/ssm-lang/ssm-runtime/actions/workflows/doc.yml/badge.svg?branch=dev"/>
</a>
<a href="https://codecov.io/gh/ssm-lang/ssm-runtime">
  <img src="https://codecov.io/gh/ssm-lang/ssm-runtime/branch/dev/graph/badge.svg?token=TYCPPY7Y92"/>
</a>
</div>

The Sparse Synchronous Model (SSM) is a deterministic real-time execution technique that allows explicit, precise timing control.

The source code for this library is hosted on GitHub at <https://github.com/ssm-lang/ssm-runtime/tree/dev>.

## Documentation

The generated documentation for this library maybe found at <https://ssm-lang.github.io/ssm-runtime>.

The code documentation is organized in terms of modules, which are listed at <https://ssm-lang.github.io/ssm-runtime/modules.html>. Note that these modules do not reflect how the documented code is organized in files. Instead, they are purely logical: they group together related families of types and symbols across C translation units and public/private interfaces.

The operation of this library was first described in:

> Stephen A. Edwards and John Hui.
> The Sparse Synchronous Model.
> In Forum on Specification and Design Languages (FDL),
> Kiel, Germany, September 2020.
> http://www.cs.columbia.edu/~sedwards/papers/edwards2020sparse.pdf

## Dependencies

The core runtime, which includes the scheduler and memory manager, is written in C99, without library dependencies.
The top-level Makefile is used to compile and test this core library, and has the following dependencies:

-   Some C99-compatible compiler (`gcc` required for testing)
-   [GNU make](https://www.gnu.org/software/make/manual/html_node/index.html)
-   [GNU Bash](https://www.gnu.org/software/bash/) (required for testing)
-   [Doxygen](https://www.doxygen.nl/index.html) (required for building documentation)
-   [Graphviz](http://graphviz.org/) (required for building documentation)
-   [valgrind](https://valgrind.org/) (required for testing)
-   [gcov](https://gcc.gnu.org/onlinedocs/gcc/Gcov.html) (required for testing)

On Ubuntu, Debian, and other Linux distributions that use aptitude, these can be installed using:

```shell
sudo apt install build-essential gcc doxygen graphviz valgrind gcov
```

Platform-specific bindings are provided via the [PlatformIO Core (CLI)](https://platformio.org) toolchain manager;
see their [installation instructions](https://docs.platformio.org/en/latest/core/installation.html).
On Linux, make sure to install the `99-platformio-udev.rules`.

Note that PlatformIO _does not use the top-level Makefile_ to build the SSM runtime.
Instead, it uses the top-level library.json manifest to compile this library as a PlatformIO package.

## Quickstart

The top-level `Makefile` can be used to build the platform-generic library alone, without platform-specific bindings.
To build just the library, just run:

```shell
make # `lib' is the default target
```

All build artifacts, including `libssm.a`, are placed in the `build` directory.

To run the included test suite, run the included test script:

```shell
./runtests.sh
```

To build and run any individual example in the `examples/` directory, e.g., `examples/fib.c`:

```shell
make build/fib
./build/fib 5
```
