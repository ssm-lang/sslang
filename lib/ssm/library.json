{
    "name": "ssm",
    "version": "0.1",
    "description": "A lightweight runtime system for the sparse synchronous model, a deterministic real-time system model",
    "repository":
    {
	"type":"git",
	"url":"https://github.com/sedwards-lab/ssm.git"
    },
    "authors":
    [
	{"name":"Stephen A. Edwards",
	 "email":"sedwards@cs.columbia.edu",
	 "url":"http://www.cs.columbia.edu/~sedwards",
	 "maintainer": true
	}
    ],
    "license":"BSD-3-Clause",
    "frameworks":"*",
    "platform":"*"
}
    
