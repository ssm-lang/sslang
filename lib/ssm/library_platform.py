Import("env")
from os.path import join, realpath

platform = env['PIOFRAMEWORK'][0]
board = env['PIOENV']

print("hello world")
print("platform: ", platform)
print("board: ", board)

platform_dir = realpath(join("platform", platform))
boards_dir = realpath(join(platform_dir, "boards"))
board_dir = realpath(join(boards_dir, board))

# Specify platform/board includes
env.Append(
    CPPPATH=[
        realpath(join("platform", platform, "include")),
        realpath(join("platform", platform, "boards", board, "include")),
    ]
)
# Specify common and platform source directories
env.Replace(
    SRC_FILTER=[
        "-<*>",
        "+<src>",
        "+<%s>" % realpath(join("platform", platform, "src")),
        "+<%s>" % realpath(join("platform", platform, "boards", board, "src")),
    ]
)
# Platform defines
env.Append(
    CPPDEFINES=[
      # This will be needed if we ever have a 64-bit timer platform,
      # to set PLATFORM_TIMER64; e.g.:
      # "PLATFORM_TIMER64",
    ]
)
