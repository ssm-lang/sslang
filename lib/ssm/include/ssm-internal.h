/** @file ssm-internal.h
 *  @brief The internal interface of the SSM runtime.
 *
 *  This header file contains definitions and declarations that should not be
 *  exposed to user code.
 *
 *  @author Stephen Edwards (sedwards-lab)
 *  @author John Hui (j-hui)
 */
#ifndef _SSM_SCHED_H
#define _SSM_SCHED_H

#include <ssm.h>

/** @ingroup error
 *  @brief Throw an internal error.
 *
 *  @platformonly
 *
 *  @param cond the condition to assert.
 */
#define SSM_ASSERT(cond)                                                       \
  do                                                                           \
    if (!(cond))                                                               \
      SSM_THROW(SSM_INTERNAL_ERROR);                                           \
  while (0)

/** @ingroup time
 *  @brief The time of the next event in the event queue.
 *
 *  Used to determine whether and when to invoke ssm_tick().
 *
 *  @platformonly
 *
 *  @returns the next event time, or #SSM_NEVER if the event queue is empty.
 */
ssm_time_t ssm_next_event_time(void);

/** @ingroup time
 *  @brief Reset the scheduler.
 *
 *  Set #now to 0; clear the event and activation record queues.
 *
 *  This does not need to be called before calling ssm_tick() for the first
 *  time; the global state automatically starts initialized.
 *
 *  @platformonly
 */
void ssm_reset(void);

/** @ingroup time
 *  @brief Advance the current model time.
 *
 *  @a next must be later than ssm_now(), and earlier than or equal to
 *  ssm_next_event_time().
 *
 *  Exposed so that platform code can perform external variable updates, to
 *  implement external inputs.
 *
 *  @platformonly
 *
 *  @param next the time to advance to.
 *
 *  @throws SSM_INVALID_TIME  @a next is earlier than or equal to #now.
 *  @throws SSM_NOT_READY     @a next is later than the earliest queued event.
 */
void ssm_set_now(ssm_time_t next);

/** @ingroup sv
 *  @brief Perform a (delayed) update on a variable.
 *
 *  Schedules all routine activation records sensitive to @a sv in the
 *  activation queue.
 *
 *  Should only be called if the variable is scheduled to be updated #now.
 *
 *  Exposed so that platform code can perform external variable updates.
 *
 *  @platformonly
 *
 *  @param sv the variable.
 *
 *  @throws SSM_NOT_READY   @a sv was not scheduled to be updated #now.
 */
void ssm_update(ssm_sv_t *sv);

/** @ingroup sv
 *  @brief Unschedule any pending events on a variable.
 *
 *  Should be called before the variable is dropped.
 *
 *  Nothing happens if the variable does not have a pending event.
 *
 *  @platformonly
 *
 *  @param var  the variable.
 */
void ssm_unschedule(ssm_sv_t *var);

/** @ingroup act
 *  @brief Run the system for the next scheduled instant.
 *
 *  If there is nothing left to run in the current instant, advance #now to the
 *  time of the earliest event in the queue, if any.
 *
 *  Removes every event at the head of the event queue scheduled for #now,
 *  updates each variable's current value from its later value, and schedules
 *  all sensitive triggers in the activation queue. Finally, execute the step
 *  functions of all scheduled routines, in order of priority.
 *
 *  Should only be called if there are activation records to execute #now,
 *  or if #now is earlier than ssm_next_event_time().
 *
 *  @platformonly
 *
 *  @throws SSM_INTERNAL_ERROR  there are stale events in the queue before #now.
 */
void ssm_tick(void);

/** @ingroup adt
 *  @brief Compute the size of a heap-allocated ADT.
 *  @todo document
 */
#define ssm_adt_size(vc) (sizeof(struct ssm_adt1) + sizeof(ssm_value_t) * ((vc) - 1))

/**
 * @addtogroup mem
 * @{
 */

/** @brief The different kinds of heap objects, enumerated.
 *
 *  Types enumerated here that are not ADTs are chosen because they cannot be
 *  easily or efficiently expressed as a product of words. For instance, 64-bit
 *  timestamps cannot be directly stored in the payload of a regular heap
 *  object, where even-numbered timestamps may be misinterpreted as pointers.
 */
enum ssm_kind {
  SSM_ADT_K = 0,  /**< ADT object, e.g., #ssm_adt1 */
  SSM_TIME_K,     /**< 64-bit timestamps, #ssm_time_t */
  SSM_SV_K,       /**< Scheduled variables, #ssm_sv_t */
};

/** @brief Initializes the underlying allocator system.
 *
 *  Memory pages are requested from the platform/OS on-demand via the provided
 *  @a alloc_page_handler. This handler must return a pointer to the beginning
 *  of a range of memory of size #SSM_MEM_PAGE_SIZE, and the page @em must be
 *  zeroed out. These pages are assumed to never be freed.
 *
 *  To support arbitrarily large allocations, SSM's allocator uses @a
 *  alloc_mem_handler to allocate memory, and @a free_mem_handler to release it.
 *  There are no requirements on the contents of memory allocated with @a
 *  alloc_mem_handler. These handlers may also assume they will not be invoked
 *  to request memory ranges of less than #SSM_MEM_POOL_MAX bytes.
 *
 *  @platformonly
 *
 *  @param alloc_page_handler   allocates pages.
 *  @param alloc_mem_handler    allocates arbitrarily large.
 *  @param free_mem_handler     frees pages allocated with @a alloc_mem_handler.
 */
void ssm_mem_init(void *(*alloc_page_handler)(void),
                  void *(*alloc_mem_handler)(size_t),
                  void (*free_mem_handler)(void *, size_t));

#ifndef SSM_MEM_POOL_MIN
/** @brief Block size of the smallest memory pool.
 *
 *  Must be strictly greater than the word size, i.e., the size of #ssm_word_t.
 *
 *  @platformonly
 */
#define SSM_MEM_POOL_MIN 16
#endif

#if SSM_MEM_POOL_MIN < SSM_POINTER_SIZE
#error SSM_MEM_POOL_MIN must be larger than word size.
#endif

#ifndef SSM_MEM_POOL_FACTOR_BASE2
/** @brief Factor between each successive memory pool size, in base 2.
 *
 *  Must be strictly greater than 0.
 *
 *  @platformonly
 */
#define SSM_MEM_POOL_FACTOR_BASE2 2
#endif

#if SSM_MEM_POOL_FACTOR_BASE2 < 1
#error SSM_MEM_POOL_FACTOR_BASE2 must be strictly greater than 0.
#endif

#ifndef SSM_MEM_POOL_COUNT
/** @brief Number of memory pools.
 *
 *  Must be strictly greater than 0.
 *
 *  @platformonly
 */
#define SSM_MEM_POOL_COUNT 4
#endif

#if SSM_MEM_POOL_COUNT < 1
#error SSM_MEM_POOL_COUNT must be strictly greater than 0.
#endif

/** @brief Compute the size of a memory pool.
 *
 *  @platformonly
 *
 *  @param pool   the 0-index of the memory pool.
 *  @returns      the size of the memory pool at position @a pool.
 */
#define SSM_MEM_POOL_SIZE(pool)                                                \
  (SSM_MEM_POOL_MIN << (SSM_MEM_POOL_FACTOR_BASE2 * pool))

/** @brief The size of the largest memory pool.
 *
 *  @platformonly
 */
#define SSM_MEM_POOL_MAX SSM_MEM_POOL_SIZE(SSM_MEM_POOL_COUNT - 1)

#ifndef SSM_MEM_PAGE_SIZE
/** @brief The size of a memory page; must be greater than #SSM_MEM_POOL_MAX.
 *
 *  @platformonly
 */
#define SSM_MEM_PAGE_SIZE SSM_MEM_POOL_SIZE(SSM_MEM_POOL_COUNT)
#endif

/** @} */

#endif /* _SSM_SCHED_H */
