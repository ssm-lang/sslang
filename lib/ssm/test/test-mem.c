#include <ssm-internal.h> /* I don't want to have to do this */
#include <ssm.h>

#include <stdio.h>

#ifndef CONFIG_MEM_STATS
#error "CONFIG_MEM_STATS must be defined to run these tests"
#endif

#ifndef SSM_NUM_PAGES
#define SSM_NUM_PAGES 32
#endif

void ssm_throw(enum ssm_error reason, const char *file, int line,
               const char *func) {
  printf("SSM error at %s:%s:%d: reason: %d\n", file, func, line, reason);
  exit(1);
}

static char mem[SSM_NUM_PAGES][SSM_MEM_PAGE_SIZE] = {0};
static size_t allocated_pages = 0;

static void *alloc_page(void) {
  if (allocated_pages >= SSM_NUM_PAGES)
    SSM_THROW(SSM_EXHAUSTED_MEMORY);
  return mem[allocated_pages++];
}

static void *alloc_mem(size_t size) { return malloc(size); }

static void free_mem(void *mem, size_t size) { free(mem); }

void print_stats(ssm_mem_statistics_t *stats) {
  ssm_mem_statistics_collect(stats);

  printf("sizeof(struct ssm_mm) = %lu\n", stats->sizeof_ssm_mm);
  printf("page size %lu\n", stats->page_size);
  printf("pages allocated %lu\n", stats->pages_allocated);
  printf("objects allocated %lu\n", stats->objects_allocated);
  printf("objects freed %lu\n", stats->objects_freed);
  printf("live objects %lu\n", stats->live_objects);

  size_t pool_count = stats->pool_count;

  printf("%lu pools\n", pool_count);

  for (size_t i = 0; i < pool_count; i++) {
    printf("pool %3lu: pages %3lu  block-size %5lu  free-blocks %5lu\n", i,
           stats->pool[i].pages_allocated, stats->pool[i].block_size,
           stats->pool[i].free_list_length);
  }

  printf("\n");
}

int main() {
  int num_errors = 0;
  ssm_mem_statistics_t stats;

  ssm_value_t objects[100];

  ssm_mem_init(alloc_page, alloc_mem, free_mem);

  printf("--- Initial state: no live objects or pages allocated\n");
  print_stats(&stats);

  objects[0] = ssm_new_blob(1);

  printf("--- One blob of size 1 allocated\n");
  print_stats(&stats);

  ssm_drop(objects[0]);

  printf("--- Back to no objects\n");
  print_stats(&stats);

  objects[0] = ssm_new_blob(1);  // pool 0
  objects[1] = ssm_new_blob(11); // pool 0
  objects[2] = ssm_new_blob(12); // pool 0

  printf("--- Three objects in pool 0\n");
  print_stats(&stats);

  objects[3] = ssm_new_blob(13); // pool 1
  objects[4] = ssm_new_blob(59); // pool 1
  objects[5] = ssm_new_blob(60); // pool 1

  printf("--- Three new objects in pool 1\n");
  print_stats(&stats);

  objects[6] = ssm_new_blob(61);  // pool 2
  objects[7] = ssm_new_blob(251); // pool 2
  objects[8] = ssm_new_blob(252); // pool 2

  printf("--- Three new objects in pool 2\n");
  print_stats(&stats);

  objects[9] = ssm_new_blob(253); // pool 3

  printf("--- One new object in pool 3\n");
  print_stats(&stats);

  ssm_drop(objects[9]);

  printf("--- Pool 3 empty\n");
  print_stats(&stats);

  ssm_drop(objects[8]);
  ssm_drop(objects[6]);
  ssm_drop(objects[7]);

  printf("--- Pool 2 empty\n");
  print_stats(&stats);

  ssm_drop(objects[0]);
  ssm_drop(objects[1]);
  ssm_drop(objects[2]);

  printf("--- Pool 0 empty\n");
  print_stats(&stats);

  ssm_drop(objects[3]);
  ssm_drop(objects[4]);
  ssm_drop(objects[5]);

  printf("--- Pool 1 empty\n");
  print_stats(&stats);

  if (num_errors == 0)
    printf("PASSED\n");
  else
    printf("FAILED (%d errors)\n", num_errors);
  return num_errors;
}
