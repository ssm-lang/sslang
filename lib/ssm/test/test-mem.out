./build/test_test-mem
# --- Initial state: no live objects or pages allocated
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 0
# objects allocated 0
# objects freed 0
# live objects 0
# 4 pools
# pool   0: pages   0  block-size    16  free-blocks     0
# pool   1: pages   0  block-size    64  free-blocks     0
# pool   2: pages   0  block-size   256  free-blocks     0
# pool   3: pages   0  block-size  1024  free-blocks     0
# 
# --- One blob of size 1 allocated
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 1
# objects allocated 1
# objects freed 0
# live objects 1
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   255
# pool   1: pages   0  block-size    64  free-blocks     0
# pool   2: pages   0  block-size   256  free-blocks     0
# pool   3: pages   0  block-size  1024  free-blocks     0
# 
# --- Back to no objects
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 1
# objects allocated 1
# objects freed 1
# live objects 0
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   256
# pool   1: pages   0  block-size    64  free-blocks     0
# pool   2: pages   0  block-size   256  free-blocks     0
# pool   3: pages   0  block-size  1024  free-blocks     0
# 
# --- Three objects in pool 0
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 1
# objects allocated 4
# objects freed 1
# live objects 3
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   253
# pool   1: pages   0  block-size    64  free-blocks     0
# pool   2: pages   0  block-size   256  free-blocks     0
# pool   3: pages   0  block-size  1024  free-blocks     0
# 
# --- Three new objects in pool 1
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 2
# objects allocated 7
# objects freed 1
# live objects 6
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   253
# pool   1: pages   1  block-size    64  free-blocks    61
# pool   2: pages   0  block-size   256  free-blocks     0
# pool   3: pages   0  block-size  1024  free-blocks     0
# 
# --- Three new objects in pool 2
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 3
# objects allocated 10
# objects freed 1
# live objects 9
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   253
# pool   1: pages   1  block-size    64  free-blocks    61
# pool   2: pages   1  block-size   256  free-blocks    13
# pool   3: pages   0  block-size  1024  free-blocks     0
# 
# --- One new object in pool 3
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 4
# objects allocated 11
# objects freed 1
# live objects 10
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   253
# pool   1: pages   1  block-size    64  free-blocks    61
# pool   2: pages   1  block-size   256  free-blocks    13
# pool   3: pages   1  block-size  1024  free-blocks     3
# 
# --- Pool 3 empty
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 4
# objects allocated 11
# objects freed 2
# live objects 9
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   253
# pool   1: pages   1  block-size    64  free-blocks    61
# pool   2: pages   1  block-size   256  free-blocks    13
# pool   3: pages   1  block-size  1024  free-blocks     4
# 
# --- Pool 2 empty
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 4
# objects allocated 11
# objects freed 5
# live objects 6
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   253
# pool   1: pages   1  block-size    64  free-blocks    61
# pool   2: pages   1  block-size   256  free-blocks    16
# pool   3: pages   1  block-size  1024  free-blocks     4
# 
# --- Pool 0 empty
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 4
# objects allocated 11
# objects freed 8
# live objects 3
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   256
# pool   1: pages   1  block-size    64  free-blocks    61
# pool   2: pages   1  block-size   256  free-blocks    16
# pool   3: pages   1  block-size  1024  free-blocks     4
# 
# --- Pool 1 empty
# sizeof(struct ssm_mm) = 4
# page size 4096
# pages allocated 4
# objects allocated 11
# objects freed 11
# live objects 0
# 4 pools
# pool   0: pages   1  block-size    16  free-blocks   256
# pool   1: pages   1  block-size    64  free-blocks    64
# pool   2: pages   1  block-size   256  free-blocks    16
# pool   3: pages   1  block-size  1024  free-blocks     4
# 
# PASSED
